### PR TITLE
feat(prowlarr): Initial tighter implementation

### LINKED ISSUE #466: Prowlarr API Key and tag support
Hi 👋🏼 

It would be awesome to instead of providing torznab urls supply a Prowlarr API key and a tag that can be used to gather the indexers.

So instead of having a config like this:

```js
module.exports = {
  qbittorrentUrl: "http://localhost",
  torznab: [
    "http://prowlarr.default.svc.cluster.local/1/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/2/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/6/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/8/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/11/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/20/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/21/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/26/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/38/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/39/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/40/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/42/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/44/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/45/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/46/api?apikey={{ .PROWLARR__API_KEY }}",
    "http://prowlarr.default.svc.cluster.local/47/api?apikey={{ .PROWLARR__API_KEY }}",
  ],
  // ...
};
```

It could be done like this:

```js
module.exports = {
  qbittorrentUrl: "http://localhost",
  prowlarrApiKey: "{{ .PROWLARR__API_KEY }}",
  prowlarrTag: "cross-seed",
  // ...
};
```

Cross seed could then look up what indexers have the `cross-seed` tag in Prowlarr and use them in their searches.

### PR DESCRIPTION
**Pull request by samip5**
Current docker image is available as `registry.samipsolutions.fi/library/cross-seed:dev-070923T1013` as the CI here is awaiting approval.

Closes #466

### CODE REVIEW

File: src/configuration.ts [hunk 1]
```diff
@@ -39,6 +39,9 @@ interface FileConfig {
 	snatchTimeout?: string;
 	searchTimeout?: string;
 	searchLimit?: number;
+	prowlarrUrl?: string;
+	prowlarrApiKey?: number;
```
Comment (onedr0p) [inline src/configuration.ts:43]: This should also be a string as well?

Comment (samip5) [reply src/configuration.ts:43]: Oh, right as it's a looooong number.. :D

File: src/prowlarr.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+		.then(fileConfig => {
+			if (fileConfig.prowlarrUrl != undefined &&
+				fileConfig.prowlarrApiKey != undefined &&
+				fileConfig.prowlarrTag != undefined) {
+
+				return axios.get(`${fileConfig.prowlarrUrl}/api/v1/indexer?apikey=${fileConfig.prowlarrApiKey}`)
+					.then(response => {
+						const indexers = response.data;
+						const filteredIndexers = indexers.filter(indexer => {
+							return indexer.tags && indexer.tags.includes(fileConfig.prowlarrTag);
```
[the remark this answers is not in the document]
Comment (onedr0p) [reply src/prowlarr.ts:15]: Good catch, after looking over the Prowlarr api docs that seems right.
https://prowlarr.com/docs/api/#/

File: src/configuration.ts (hunk 1 above)
Comment (onedr0p) [reply src/configuration.ts:43]: I think it can be all number but it can also have `[a-z]` in it as well.

**Comment by onedr0p** [comment]
I was chatting with @samip5 and hopefully this solution of populating the torznab list on start with the Prowlarr indexers is fine. There is probably other ways to tackle this but this seemed like the easiest.

**Review by buroa — COMMENTED** [review]
Besides the `prowlarrTag` stuff @mynameisbogdan mentioned, this looks really good and will be a awesome feature. Would also request some live reloading of this, incase on Prowlarr side things change, we don't need to roll the application.

**Comment by samip5** [comment]
> Would also request some live reloading of this, incase on Prowlarr side things change, we don't need to roll the application.

How do you imagine that would work? Some interval that it checks if things have changed?

**Comment by samip5** [comment]
> It seems like the solution needs to be reworked to not populate the torznab config object and instead just pull directly from Prowlarr on a search if the settings are defined. Cross seed should check in Prowlarr that the indexer has the tag, that is type `torrent`, that it is not disabled and that Prowlarr hasn't marked it as down (e.g. the site is down).

That would require much more due to how cross-seed is dependent on database, which is why I opted for this approach.

**Comment by buroa** [comment]
> > Would also request some live reloading of this, incase on Prowlarr side things change, we don't need to roll the application.
> 
> How do you imagine that would work? Some interval that it checks if things have changed?

Either a timer inside the app... or... a little bit of a harder way but maybe better down the road: cross seed supports a new API to accept indexers from Prowlarr directly. Just like how we sync them to Sonarr/Radarr. That would require additional engineering efforts on Prowlarr as well, though.

**Comment by onedr0p** [comment]
Yeah I think this is a good start, it just means we have to bounce the app anytime we change tags in Prowlarr.

**Comment by samip5** [comment]
Okay, the current code that I pushed is available as `registry.samipsolutions.fi/library/cross-seed:dev-310823T0007`

**Comment by samip5** [comment]
`registry.samipsolutions.fi/library/cross-seed:dev-310823T0052`

**Comment by onedr0p** [comment]
Looks like it is working!

```
✖ k exec --stdin --tty -n default cross-seed-7c7974c6d5-lhwxb -- cross-seed search --no-include-episodes
Configuration file already exists.
Configuration file already exists.
info: Validating your configuration...
info: Your configuration is valid!
info: [search] Found 538 torrents, 185 suitable to search for matches
info: [1/185] Searching for Awesome.Show.S03.1080p.AMZN.WEB-DL.DD+2.0.H.264-NOGROUP
info: [torznab] Searching 18 indexers for Curb.Your.Enthusiasm.S03.1080p.AMZN.WEB-DL.DD+2.0.H.264-AJP69
```

**Comment by samip5** [comment]
> @samip5 how do you pull this, I was using `ghcr.io/autobrr/samip5:prowlarr-integration` as the repo?

Bu pulling the image as `registry.samipsolutions.fi/library/cross-seed:dev-310823T0101`

**Comment by mmgoodnow** [comment]
> > It seems like the solution needs to be reworked to not populate the torznab config object and instead just pull directly from Prowlarr on a search if the settings are defined. Cross seed should check in Prowlarr that the indexer has the tag, that is type `torrent`, that it is not disabled and that Prowlarr hasn't marked it as down (e.g. the site is down).
> 
> That would require much more due to how cross-seed is dependent on database queries, which is why I opted for this approach.
> 
> Related: https://github.com/cross-seed/cross-seed/blob/308fad8aaa4871994bd862451512e25c5096d724/src/indexers.ts#L29L49

one potential solution would be to make the body of `getEnabledIndexers` to hit the Prowlarr api, though it might have to be cached if it gets called in the hot path. I don't really remember…

File: src/prowlarr.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+export async function getProwlarrIndexers(): Promise<string[] | void> {
+	try {
+		const fileConfig = await getFileConfig();
+		if (fileConfig.prowlarrUrl !== undefined &&
+			fileConfig.prowlarrApiKey !== undefined &&
+			fileConfig.prowlarrTag !== undefined) {
+
+			try {
+				const tagId = await fetchTagId(fileConfig);
+				const response = await axios.get(`${fileConfig.prowlarrUrl}/api/v1/indexer?apikey=${fileConfig.prowlarrApiKey}`);
```
[the remark this answers is not in the document]
Comment (samip5) [reply src/prowlarr.ts]: That's a good catch, it definitely should be inside the null check.

**Comment by fabricionaweb** [comment]
is it axios really necessary?

**Comment by samip5** [comment]
> is it axios really necessary?

What's wrong with axios?

**Comment by onedr0p** [comment]
I don't think anything is wrong with it, but this app is using `node-fetch` in other places instead so it might be good to consolidate on that since it's already in use.

**Review by onedr0p — APPROVED** [review]

**Review by buroa — APPROVED** [review]

**Comment by KyleSanderson** [comment]
Kind of wild this is going-on when Jackett was removed well over a year ago. https://github.com/cross-seed/cross-seed/commit/6e8610427443652c0780b4b2ac4ebf09d2be86d0#diff-57ac2b87741482cfb4870debdf993b6ada662a9830056678cb3f7d70dfad0a38L45

It's a good addition, and the removal never made sense, but I'm sensing a bit of heartbreak as there was no communication (or project history even) considered with this.

**Comment by samip5** [comment]
> It's a good addition, and the removal never made sense, but I'm sensing a bit of heartbreak as there was no communication (or project history even) considered with this.

What do you mean by that exactly?

**Comment by mmgoodnow** [comment]
He's saying I am likely not going to merge it. 

And he's not exactly wrong - I am incredibly wary of maintenance surface because I have pretty limited time to devote to cross-seed. so something like this, which will become tech debt (requiring a breaking change to remove) when I want to move to a world where indexers are database driven rather than config driven, is a bit of a hard sell.

That said, I appreciate the effort here, and I want to find a resolution that works for everyone here (it might involve starting the charge to make indexers database-driven). @samip5 would you mind reaching out to me on discord?

**Comment by samip5** [comment]
> I want to move to a world where indexers are database driven rather than config driven, is a bit of a hard sell.

That's not a very stateless direction IMO, and I for one are against that so this seems to potentially be a time for a fork that doesn't have a database at all, and instead tight Prowlarr integration.

> That said, I appreciate the effort, and I want to find a resolution that works for everyone here (it might involve starting the charge to make indexers database-driven).

I'm not sure if that can be achieved if your goal is to make it even more database driven.

**Comment by KyleSanderson** [comment]
> > I want to move to a world where indexers are database driven rather than config driven, is a bit of a hard sell.
> 
> That's not a very stateless direction IMO, and I for one are against that so this seems to potentially be a time for a fork that doesn't have a database at all, and instead tight Prowlarr integration.

Sorry, you're saying you're not preserving the torrent cache, and are thus endlessly downloading files from trackers? how are you not disabled on every tracker yet. You should absolutely be preserving all state from this application across runs - anything else is abusive and reckless to the community.

> > That said, I appreciate the effort, and I want to find a resolution that works for everyone here (it might involve starting the charge to make indexers database-driven).
> 
> I'm not sure if that can be achieved if your goal is to make it even more database driven.

Of course it can. As part of your init container you can populate any local database, as you would normally do in any enterprise environment to bootstrap the pod. That's literally why the entire mechanism exists in k8s.

This isn't my project (I have no role here at all, really). I've seen the group of you move around from project to project to try and get your way by proposing crazy things (the ask, isn't to me, but the "reason" is flawed), but I would highly encourage you to have a more open perspective to the realities of today, and tomorrow outside of the echo chamber.

**Comment by onedr0p** [comment]
> Sorry, you're saying you're not preserving the torrent cache, and are thus endlessly downloading files from trackers? how are you not disabled on every tracker yet. You should absolutely be preserving all state from this application across runs - anything else is abusive and reckless to the community.

I only want to cross seed on completion of a torrent or when autobrr finds an announcement, I don't want to endlessly run cross-seed on an interval to search trackers. You don't need state for this.

> Of course it can. As part of your init container you can populate any local database, as you would normally do in any enterprise environment to bootstrap the pod. That's literally why the entire mechanism exists in k8s.

What are you even trying to solve here? Cross seed uses a sqlite database why would you need an init container to create and populate a sqlite database when the application does it?

> This isn't my project (I have no role here at all, really). I've seen the group of you move around from project to project to try and get your way by proposing crazy things (the ask, isn't to me, but the "reason" is flawed), but I would highly encourage you to have a more open perspective to the realities of today, and tomorrow outside of the echo chamber.

Again your attitude and arrogance keeps getting the best of you, understand that the point of open source is to collaborate, improve, iterate and share ideas. If you do not want to do those things, please stick to close sourced software.

**Comment by samip5** [comment]
> Sorry, you're saying you're not preserving the torrent cache, and are thus endlessly downloading files from trackers?

There are other ways to achieve that than a sqlite database.

> This isn't my project (I have no role here at all, really). I've seen the group of you move around from project to project to try and get your way by proposing crazy things (the ask, isn't to me, but the "reason" is flawed), but I would highly encourage you to have a more open perspective to the realities of today, and tomorrow outside of the echo chamber.

I wouldn't call this crazy tbh, but please explain how the reason is flawed to you?

**Comment by samip5** [comment]
I will not contribute to this project if Kyle remains a active admin on this project.

**Comment by KyleSanderson** [comment]
> I will not contribute to this project if Kyle remains an active admin on this project.

you're a nut case. Again, as stated, I have no privileges or direction here. I'm another user, just like yourself. If you're calling for a ban, I'm okay with that outcome as well.

I highly suggest some maturity here.

**Comment by samip5** [comment]
> > I will not contribute to this project if Kyle remains an active admin on this project.
> 
> you're a nut case. Again, as stated, I have no privileges or direction here. I'm another user, just like yourself. If you're calling for a ban, I'm okay with that outcome as well.
> 
> I highly suggest some maturity here.

Please speak for yourself.

**Comment by bakerboy448** [comment]
@samip5 i wouldn't let kyle's general childish assholery and wandering commentary scare you off from contributing....that's his personality and it seems like he doesn't get the hint he should stick to code and avoid interacting with people...especially with how hostile his commentary often is...

**Comment by Motophan** [comment]
I'm w/ kyle on this one. 
Sane defaults, not spamming trackers over and over and over w/ downloads leading people to get disabled for actual abuse. 

If you need to have abusive behavior in regards to tech debt regarding this app I will use the kyle-no-abuse fork.

**Comment by bakerboy448** [comment]
Got reports of abuse? Doesn't seem there's been any?

**Comment by Motophan** [comment]
Spam downloading the same .torrent is abusive and will get you banned on any non-public tracker. It will eventually get you rate limited on public ones aswell. Kyle is 100% right in this regard and I support Kyle going forward. 

If you want, feel free to run a script to download the .torrent every 1 second to /dev/null and lets see how long your account lasts. 

`#!/bin/bash
while true
do
    curl -s "https://example.com/torrent.php?action=download&id=1234&torrent_pass=1234" > /dev/null
    sleep 1
done`

**Comment by bakerboy448** [comment]
Again, what does this have to do with tighter prowlarr integration....please take your Kyle-like behavior of hostility and trolling on unrelated PRs elsewhere or go to discord to discuss the "abuse" or open a bug report of the "abuse" with proper verbose logging demonstrating the issue

**Comment by onedr0p** [comment]
@Motophan makes sense but this has nothing to do with this PR, it's not even a concern with the modifications that were made. Please stop getting offtopic.

**Pull request closed without merging**